### PR TITLE
Install script update, adding configuration and possibility to have a private instance

### PR DESCRIPTION
**Pull request by rokaz**
Hi,

I'm pushing this one, it is pretty big, I'm sorry that I didn't break it up into separate pieces, if necessary I can do it later.

### CODE REVIEW

**Comment by lapineige** [comment]
What a big PR ^^
Thanks a lot, after a quick look it seems to be ok. I'll test installing it :)

File: scripts/install [hunk 1]
```diff
@@ -29,53 +52,70 @@ fi
...
-sudo chown -R www-data: $app_home_path
-
+# generate the salt
 salt=$(dd if=/dev/urandom bs=1 count=200 2> /dev/null | tr -c -d 'A-Za-z0-9' | sed -n 's/\(.\{24\}\).*/\1/p')
+
+# generate the hash with the password
+password_hash=$(php ../conf/gen_hash.php $password $admin $salt)
+
+# set the proper value for substitution
 if [ $privatelinkbydefault = "No" ]; then
```
Comment (lapineige) [inline scripts/install]: If we setup a private instance, I don't think this option is needed.

File: manifest.json [hunk 2]
```diff
@@ -10,20 +13,49 @@
...
+        "en": "Owner of the Shaarli instance",
+        "fr": "Propriétaire de l'instance Shaarli"
+      },
+      "example": "test"
+    },
+    {
+      "name": "password",
+      "type": "password",
+      "ask": {
+        "en": "Define password for Shaarli user",
```
Comment (lapineige) [inline manifest.json]: As it's useless for a private instance, maybe we can tell the user that s.he can live it blank in that case ?

Comment (lapineige) [inline manifest.json]: Ok giving a blank password is not possible. So we can just tell the user to use a random password as it is not required.

**Comment by lapineige** [comment]
General question: I'm wondering if we should adapt the app name (I mean, the label), as multiple can be installed.
So the administrator might be less confused, and if the user decide to install a private + a public instance (which is possible, isn't it ?), s.he will never mistake the apps.

**Comment by lapineige** [comment]
Install, backup and restore are working.
Tested with single user Yunohost and 2 Shaarli private installations.

I did the same for a public install, and it's impossible to login.
---
In all cases, this message is raised:
> Attention : PHP Notice:  Undefined offset: 3 in /var/cache/yunohost/from_file/shaarli_ynh-install_script/conf/gen_hash.php on line 4

But I didn't figured out what's the issue, and saw no errors when using Shaarli.

File: scripts/install (hunk 1 above)
Comment (rokaz) [inline scripts/install]: Yes, but we leave the possibility to have a public instance, and I don't know if it is possible not to set the password in the config file. Anyway, as we always ask for the password (we don't have the possibility to have conditions in the manifest), we should set this anyway IMHO

**Comment by rokaz** [comment]
I don't know if it is possible to change the name of the instance, or to label it, we should ask the yunohost devs. 
Regarding the error you get, I don't know, if I remember correctly I also tested the app as a public site and didn't have problems to log in. I will check next week.

File: scripts/install (hunk 1 above)
Comment (lapineige) [inline scripts/install]: You're right, no real to skip this step as we have to prompt the user for a password anyway.
Let's keep this simple.

**Comment by lapineige** [comment]
> I don't know if it is possible to change the name of the instance, or to label it, we should ask the yunohost devs.

You can, for instance with:
>sudo yunohost app setting shaarli label -v "aNewLabel"
sudo yunohost app ssowatconf

For instance I did it when migrating from Owncloud to Nextcloud. It worked like a charm.

**Comment by rokaz** [comment]
I tested the installation of a public site and I don't have any problem to log in, either if I am logged in as a Yunohost user (owner of the instance or not) or not logged in at all.

After looking at it, the error raised occurred during the install: the salt was not generated (exhaustion of /dev/urandom on your machine? maybe we should generate the salt with a different method?) and thus the hash generation failed. Could you check that the salt field in /var/www/$shaarli/data/config.json.php is empty? It should be empty on your system after this error and should not be on a working system. This would explain the impossibility to log in.

**Pull request merged**